### PR TITLE
Move impImportCall to a separate class.

### PR DESCRIPTION
**Pull request by sandreenko**
We regularly make changes in Compiler::impImportCall, for example for CoreRT. 
This function was ~1500 lines and it was hard to change it.
During this change some strange code was found, like unreachable paths or always true conditions, that will be  fixed in next PRs.

### CODE REVIEW

**Comment by sandreenko** [comment]
Could somebody please take a look @dotnet/jit-contrib ?

**Comment by sandreenko** [comment]
No spmi diffs.

**Comment by sandreenko** [comment]
@dotnet-bot test Tizen armel Cross Release Build 
@dotnet-bot test Ubuntu arm Cross Release Build

**Comment by sandreenko** [comment]
@dotnet-bot test Tizen armel Cross Release Build

**Comment by BruceForstall** [comment]
Please do full (all platform) desktop testing (including asm diffs). Also, do stress modes + R2R testing.

File: src/jit/importcall.cpp [hunk 1]
```diff
@@ -0,0 +1,1714 @@
...
+XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+*/
+class CallImporter
+{
+public:
+    //------------------------------------------------------------------------
+    // CallImporter constructor
+    //
+    // Arguments:
+    //    compiler
```
Comment (BruceForstall) [inline src/jit/importcall.cpp]: missing a few :-)

File: src/jit/importcall.cpp [hunk 2]
```diff
@@ -0,0 +1,1714 @@
...
+        intrinsicID = CORINFO_INTRINSIC_Count;
+    }
+
+    //------------------------------------------------------------------------
+    // impImportCall: see the Compiler::impImportCall description.
+    // All parameters are initialized in the contructor.
+    //
+    // Returns:
+    //    Type of the call's return value.
+    var_types importCall()
```
Comment (BruceForstall) [inline src/jit/importcall.cpp]: https://github.com/dotnet/coreclr/blob/master/Documentation/coding-guidelines/clr-jit-coding-conventions.md#8.9

File: src/jit/importcall.cpp [hunk 3]
```diff
@@ -0,0 +1,1714 @@
...
+        return callRetTyp;
+    }
+
+private:
+    //------------------------------------------------------------------------
+    // checkInlineRestrictions: Check does this call satisfy inlining restrictions.
+    //
+    // Return Value:
+    //    true if does, false if not.
+    bool checkInlineRestrictions()
```
Comment (BruceForstall) [inline src/jit/importcall.cpp]: name it `CallSatisfiesInlineRestrictions()`? or `CallCanBeInlined()`?

File: src/jit/importcall.cpp [hunk 4]
```diff
@@ -0,0 +1,1714 @@
...
+                    // TODO: Still using the widened type.
+                    call = compiler->gtNewInlineCandidateReturnExpr(call, genActualType(callRetTyp));
+                }
+                else
+                {
+                    if (fatPointerCandidate)
+                    {
+                        // fatPointer candidates should be in statements of the form call() or var = call().
+                        // Such form allows to find statements with fat calls without walking through whole trees
+                        // and removes problems with
```
Comment (BruceForstall) [inline src/jit/importcall.cpp]: It would be useful if these member variable declarations (eventually) had comments. I realize you copy-and-pasted them, so no need to do it now.

Are they all required to be member vars? Are all of them used by at least 2 different functions?

**Comment by sandreenko** [comment]
@dotnet-bot test Windows_NT x86 Checked tailcallstress
@dotnet-bot test Windows_NT x86 Checked r2r_jitstressregs2
@dotnet-bot test Windows_NT jitstress2
@dotnet-bot test Windows_NT x64 corefx_jitstress2
@dotnet-bot test Windows_NT Release pri1r2r

**Comment by AndyAyersMS** [comment]
Before we get too deep on the particulars,  I am very much on the fence about this kind of refactoring. 

On the one hand I appreciate that the method is big and complicated, and changing it is not easy, and something probably should be done. 

On the other hand I'm not convinced that this change is an improvement, or is attacking the essential problems, or is even the right place to expend refactoring calories.

If we are going to do this sort of thing, I'd rather see some up-front discussions and consensus on what exactly we should change and how we should go about effecting change, rather than to try and work back to that point from a big pile of changes.

**Comment by sandreenko** [comment]
> On the other hand I'm not convinced that this change is an improvement, or is attacking the essential problems, or is even the right place to expend refactoring calories.

I regularly change this method for CoreRT issues. With each change (e.g. #9191) it becomes bigger and bigger. I already spent a lot of time reading this and I want to optimize in for the future,

**Comment by AndyAyersMS** [comment]
Can you summarize what you've learned from your experiences making changes to this method? What do you see as the advantages of your new approach? What are the risks? What alternatives did you consider and decide were lacking?

**Comment by sandreenko** [comment]
> Can you summarize what you've learned from your experiences making changes to this method?

I can summarize that such changes do not take too long to do, but uncover bad or outdated code and comments. The worst thing was to delete `goto ` without too much new code. Another problem was that this function had too many exit points and before I understand that we don't use the result of the call for failed cases ([return TYP_UNDEF instead of callReturnType for failed cases. ](https://github.com/dotnet/coreclr/pull/13091/commits/0322d26ffdb42dfb2ce05fce50215fd0ad4c691c)) it was hard to track them all.

> What do you see as the advantages of your new approach?

It allows to do new changes cheaper and faster. It allows to track changes in this function better. It uncovered some strange code ([delete unreached code](https://github.com/dotnet/coreclr/pull/13091/commits/5ca6837add31aad4f244bae5d450214caf0a7f9e) and [extract getPInvokeCookie ](https://github.com/dotnet/coreclr/pull/13091/commits/c2938163070baf0695128b591b72ca29991419d7)commits).
So if it is merged, I will create some PRs that delete unreached paths or always true expression. It will be easier to review them after this one.

> What are the risks?

Because there is no spmi diffs (and should be) and this function is very hot and well covered in our tests, I do not see many risks. I will do risky changes in a separate PR.

> What alternatives did you consider and decide were lacking?

The main alternative was to separate this function into 4 for each call opcode (CEE_NEWOBJ, CEE_CALL, CEE_CALLI, CEE_CALLVIRT), because many arguments are opcode specific. But after a try I understood that it would be better to do it after this PR.

File: src/jit/importcall.cpp (hunk 3 above)
Comment (sandreenko) [inline src/jit/importcall.cpp]: Fixed

File: src/jit/importcall.cpp (hunk 2 above)
Comment (sandreenko) [inline src/jit/importcall.cpp]: Fixed

**Comment by sandreenko** [comment]
@dotnet-bot test Windows_NT x86 Checked tailcallstress
@dotnet-bot test Windows_NT x86 Checked r2r_jitstressregs2
@dotnet-bot test Windows_NT jitstress2
@dotnet-bot test Windows_NT x64 corefx_jitstress2
@dotnet-bot test Windows_NT Release pri1r2r

**Comment by sandreenko** [comment]
@dotnet-bot test Windows_NT x86 Checked tailcallstress
@dotnet-bot test Windows_NT x86 Checked r2r_jitstressregs2
@dotnet-bot test Windows_NT jitstress2
@dotnet-bot test Windows_NT x64 corefx_jitstress2
@dotnet-bot test Windows_NT Release pri1r2r

**Comment by sandreenko** [comment]
PR was updated.
Please take another look.

**Comment by sandreenko** [comment]
@dotnet-bot test Tizen armel Cross Release Build

**Comment by sandreenko** [comment]
No spmi diffs.

**Comment by sandreenko** [comment]
@dotnet-bot test Tizen armel Cross Debug Build

**Comment by karelz** [comment]
@sandreenko what is status of this PR? There was no update for last 9 days ...

**Comment by sandreenko** [comment]
Looks like this PR is not wanted to be merged by some reasons.

Maybe things will change in the future or I will be able to merge this changes in separate small PRs.
